plugins/usbpro/OpenDeckWidget.cpp

```diff
@@ -0,0 +1,79 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * OpenDeckWidget.cpp
+ * The OpenDeck Widget.
+ * Copyright (C) 2021 Simon Newton, Igor Petrovic
+ */
+
+#include <vector>
+
+#include "ola/Constants.h"
+#include "plugins/usbpro/OpenDeckWidget.h"
+
+namespace ola {
+namespace plugin {
+namespace usbpro {
+
+/**
+ * OpenDeckWidget Constructor
+ */
+OpenDeckWidget::OpenDeckWidget(
+  ola::io::ConnectedDescriptor *descriptor)
+    : GenericUsbProWidget(descriptor) {
+}
+
+
+bool OpenDeckWidget::SendDMX(const DmxBuffer &data) {
+  if (data != internal_buffer) {
+    std::vector<uint8_t> send_buffer = {};
+
+    for (size_t index = 0; index < data.Size(); index++) {
+      if (data.Get(index) != internal_buffer.Get(index)) {
+        send_buffer.push_back((index+1) & 0xFF);
+        send_buffer.push_back((index+1) >> 8);
+        send_buffer.push_back(data.Get(index));
+      }
+    }
+
+    internal_buffer = data;
+
+    if (send_buffer.size() >= MAX_DIFF_CHANNELS) {
```

Comment:
Perhaps put this check inside the for loop too, otherwise for a blackout it will generate the full send_buffer first, then realise it doesn't need to.

Also either this calculation is wrong, or the variable name is vague, if the MAX_DIFF_CHANNELS is 3, then it will only be one actual channel, comprising of it's high byte, low byte and value.

Reply:
You're right. I've reworked this entire function so please take another look.

Reply:
This looks much better now thanks!